core/src/main/scala/kafka/log/AbstractLazyIndex.scala

```diff
@@ -0,0 +1,85 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package kafka.log
+
+import java.io.File
+import java.util.concurrent.locks.ReentrantLock
+
+import kafka.utils.CoreUtils.inLock
+
+/**
+ * An abstract lazy index wrapper that provides capability of avoiding index initialization
+ * on construction.
+ *
+ * This defers the index initialization to the time it gets accessed
+ * so the cost of the heavy memory mapped operation gets amortized over time.
+ */
+abstract class AbstractLazyIndex[T <: AbstractIndex[_, _]](private val _file: File) {
```

Comment:
I am wondering if we should keep this in its own file, or move-in with `AbstractIndex.scala` since both `LazyTimeIndex` and `LazyOffsetIndex` are kept together with `TimeIndex` and `OffsetIndex` definitions. Alternatively we can move them out into their own files. WDYT @ijuma?

Reply:
Makes sense to keep it in the same file as the `Lazy*Index` implementations.

Reply:
Oh, I see. The current structure is a bit weird. I think I'd keep all the Lazy implementations together. If you don't mind, I'll push a commit to your branch and we can keep/remove/adjust after.

Reply:
sounds good!